### PR TITLE
Feature/date formatting

### PR DESCRIPTION
**Pull request by ganfra**
This PR takes care of centralizing most of the date formating in the app. 
It also adds support for am/pm time display, and fix year not being shown when needed

### CODE REVIEW

**Comment by bmarty** [comment]
@ganfra Should be nice to tell which issue is closed by this PR (maybe but not limited to this list: https://github.com/vector-im/element-android/labels/feature%3Adate_time)

File: vector/src/main/java/im/vector/app/core/date/AbbrevDateFormatterProvider.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+class AbbrevDateFormatterProvider @Inject constructor(private val localeProvider: LocaleProvider) : DateFormatterProvider {
+
+    override val dateWithMonthFormatter: DateTimeFormatter by lazy {
+        val pattern = DateFormat.getBestDateTimePattern(localeProvider.current(), "d MMM")
+        DateTimeFormatter.ofPattern(pattern, localeProvider.current())
+    }
+
+    override val dateWithYearFormatter: DateTimeFormatter by lazy {
+        val pattern = DateFormat.getBestDateTimePattern(localeProvider.current(), "dd.MM.yyyy")
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/date/AbbrevDateFormatterProvider.kt:32]: I think English people are happy with `MM.dd.yyyy` as strange as it looks

File: vector/src/main/java/im/vector/app/core/date/DefaultDateFormatterProvider.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import android.text.format.DateFormat
+import im.vector.app.core.resources.LocaleProvider
+import org.threeten.bp.format.DateTimeFormatter
+import javax.inject.Inject
+
+class DefaultDateFormatterProvider @Inject constructor(private val context: Context,
+                                                       private val localeProvider: LocaleProvider)
+    : DateFormatterProvider {
+
+    override val dateWithMonthFormatter: DateTimeFormatter by lazy {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/date/DefaultDateFormatterProvider.kt:29]: Should be nice to add example of formatted date in the Javadoc of all this methods, or maybe document the enum `DateFormatKind` please

File: vector/src/main/java/im/vector/app/core/date/VectorDateFormatter.kt [hunk 3]
```diff
@@ -19,64 +19,147 @@ package im.vector.app.core.date
...
 import org.threeten.bp.format.DateTimeFormatter
-import java.util.Calendar
-import java.util.Date
 import javax.inject.Inject
-
-/**
- * Returns the timestamp for the start of the day of the provided time.
- * For example, for the time "Jul 21, 11:11" the start of the day: "Jul 21, 00:00" is returned.
- */
-fun startOfDay(time: Long): Long {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/date/VectorDateFormatter.kt:33]: This was added by a PR from the community: https://github.com/vector-im/element-android/pull/1803 to fix #822 

Is it OK to remove it?

**Review by bmarty — COMMENTED** [review]
LGTM, 3 remark though

File: vector/src/main/java/im/vector/app/core/date/AbbrevDateFormatterProvider.kt (hunk 1 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/core/date/AbbrevDateFormatterProvider.kt:32]: Yes it's handled by the method getBestPattern

File: vector/src/main/java/im/vector/app/core/date/DefaultDateFormatterProvider.kt (hunk 2 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/core/date/DefaultDateFormatterProvider.kt:29]: Yes

File: vector/src/main/java/im/vector/app/core/date/VectorDateFormatter.kt (hunk 3 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/core/date/VectorDateFormatter.kt:33]: Yes, I don't use the relative method anymore

**Comment by bmarty** [comment]
Thanks for the update. ktlint is not happy

**Comment by bmarty** [comment]
Rebased, will be merged once CI is happy

**Review by bmarty — APPROVED** [review]
LGTM, thanks!

**Pull request merged**